pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go

```diff
@@ -208,14 +212,162 @@ func startDistIngestion(
+		ju.UpdateProgress(md.Progress)
+		return nil
+	}); err != nil {
+		return err
+	}
+
 	err = ctxgroup.GoAndWait(ctx, execInitialPlan, replanner, tracingAggLoop, streamSpanConfigs)
 	if errors.Is(err, sql.ErrPlanChanged) {
 		execCtx.ExecCfg().JobRegistry.MetricsStruct().StreamIngest.(*Metrics).ReplanCount.Inc(1)
 	}
 	return err
 }
 
+// TODO(ssd): This is a duplicative with the split_and_scatter processor in
+// backupccl.
+type splitAndScatterer interface {
+	split(
+		ctx context.Context,
+		splitKey roachpb.Key,
+		expirationTime hlc.Timestamp,
+	) error
+
+	scatter(
+		ctx context.Context,
+		scatterKey roachpb.Key,
+	) error
+
+	now() hlc.Timestamp
+}
+
+type dbSplitAndScatter struct {
+	db *kv.DB
+}
+
+func (s *dbSplitAndScatter) split(
+	ctx context.Context, splitKey roachpb.Key, expirationTime hlc.Timestamp,
+) error {
+	return s.db.AdminSplit(ctx, splitKey, expirationTime)
+}
+
+func (s *dbSplitAndScatter) scatter(ctx context.Context, scatterKey roachpb.Key) error {
+	_, pErr := kv.SendWrapped(ctx, s.db.NonTransactionalSender(), &kvpb.AdminScatterRequest{
+		RequestHeader: kvpb.RequestHeaderFromSpan(roachpb.Span{
+			Key:    scatterKey,
+			EndKey: scatterKey.Next(),
+		}),
+		RandomizeLeases: true,
+		MaxSize:         1, // don't scatter non-empty ranges on resume.
+	})
+	return pErr.GoError()
+}
+
+func (s *dbSplitAndScatter) now() hlc.Timestamp {
+	return s.db.Clock().Now()
+}
+
+// createInitialSplits creates splits based on the given toplogy from the
+// source.
+//
+// The idea here is to use the information from the source cluster about
+// the distribution of the data to produce split points to help prevent
+// ingestion processors from pushing data into the same ranges during
+// the initial scan.
+func createInitialSplits(
+	ctx context.Context,
+	codec keys.SQLCodec,
+	splitter splitAndScatterer,
+	topology streamclient.Topology,
+	destTenantID roachpb.TenantID,
+) error {
+	ctx, sp := tracing.ChildSpan(ctx, "streamingest.createInitialSplits")
+	defer sp.Finish()
+
+	rekeyer, err := backupccl.MakeKeyRewriterFromRekeys(codec,
+		nil /* tableRekeys */, []execinfrapb.TenantRekey{
+			{
+				OldID: topology.SourceTenantID,
+				NewID: destTenantID,
+			}},
+		true /* restoreTenantFromStream */)
+	if err != nil {
+		return err
+	}
+	for _, partition := range topology.Partitions {
+		for _, span := range partition.Spans {
+			startKey := span.Key.Clone()
+			splitKey, _, err := rekeyer.RewriteKey(startKey, 0 /* walltimeForImportElision */)
+			if err != nil {
+				return err
+			}
+
+			// NOTE(ssd): EnsureSafeSplitKey called on an arbitrary
+			// key unfortunately results in many of our split keys
+			// mapping to the same key for workloads like TPCC where
+			// the schema of the table includes integers that will
+			// get erroneously treated as the column family length.
+			//
+			// Since the partitions are generated from a call to
+			// PartitionSpans on the source cluster, they should be
```

Comment:
I wonder if it makes sense for the source cluster to create an iterator and read then key (SeekGE) that's at or after the split point? This could guarantee that the explicit split keys are valid.

Reply:
Just to make sure I'm following.

This would look like taking the output of PartitionSpans. Then iterate over the results. For each start key of the spans returned by PartitionSpans, we SeekGE to that key and return the first key rather than the split point generated by PartitionSpan.  

The pro here is that the key returned is definitely going to be a SQL key that we can run back through EnsureSafeSplitKey.

The cons here would be the bit of added complexity on the source planning side and also that, at least for some time after the initial scan, on the destination you'd have splits that were just a little off from the source. So if a lot of data gets put "in between" the start according to partition spans and the actual start key, that data would now go to the lhs range.  During the initial scan this shouldn't matter because we do our scan at a fixed timestamp that should be rather close to when we produced this original plan.

Reply:
Yup; that about sums it up.  I was just trying to come up w/ a way to avoid this "scary" warning you wrote re not calling ensure safe key.